src/bin/stratisd.rs

```diff
@@ -85,20 +85,45 @@ fn run() -> StratisResult<()> {
 
     let (dbus_conn, mut tree, dbus_context) = libstratis::dbus_api::connect(Rc::clone(&engine))?;
 
-    // Get a list of fds to poll for
-    let mut fds: Vec<_> = dbus_conn
-        .watch_fds()
-        .iter()
-        .map(|w| w.to_pollfd())
-        .collect();
+    let mut fds = Vec::new();
+
+    let mut eventable = engine.borrow_mut().get_eventable()?;
+    let dbus_fd_start_idx = match eventable {
+        Some(ref mut evt) => {
+            fds.push(libc::pollfd {
+                         fd: evt.get_pollable_fd(),
+                         revents: 0,
+                         events: libc::POLLIN,
+                     });
+            1
+        }
+        None => 0,
+    };
+
+    let poll_timeout = if dbus_fd_start_idx == 0 { 10000 } else { -1 };
 
     loop {
-        // Poll them with a 10 s timeout
-        let r = unsafe { libc::poll(fds.as_mut_ptr(), fds.len() as libc::c_ulong, 10000) };
-        assert!(r >= 0);
+        // Handle the Eventable (if there is one) before polling. If an event
+        // arrives after calling clear_event(), we will pop out of the poll
```

Comment:
most of this comment is obscure to me. "If an event arrives after calling clear_event()``` is the unclearest to me. I'm not sure we do guarantee that no events are missed. We could miss 100 and then detect the 101st, AFAICT.

Reply:
Can you give an example of scenario where you think we would miss an event? Then I can walk through it.

Reply:
Well I'm assuming a one-to-one and onto relationship between events and event nrs. Suppose you call list devices. You look at all devices and the highest event_nr is n. Later you call list_devices again. The device which previously had event_nr n now has event_nr n + 2. All other devices had the same event nr they had previously. I don't think that is impossible, but that means that you missed event corresponding to event nr n + 1.

Reply:
Yes, that is certainly possible. That's working as intended. The goal is to get woken up when an event happens, and then know which DM device(s) evented (once or more) times since the last time we handled events. The assumption is we'd get the current status of the device and take action based on the status. If the status shows multiple things needing attention with the device then we should resolve all of them, rather than getting a separate event for each issue.